presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +76,72 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<ConnectorSplitSource> future;
+
+        private KuduDynamicFilteringSplitSource(
+                ConnectorTableHandle table,
+                SplitSchedulingStrategy splitSchedulingStrategy,
+                DynamicFilter dynamicFilter,
+                long blockingTimeoutMillis)
+        {
+            this.future = whenCompleted(dynamicFilter)
+                    .completeOnTimeout(null, blockingTimeoutMillis, MILLISECONDS)
+                    .thenApply(o -> getSplitSource(table, splitSchedulingStrategy, dynamicFilter));
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+        {
+            return future.thenCompose(s -> s.getNextBatch(partitionHandle, maxSize));
+        }
+
+        @Override
+        public void close()
+        {
+            if (future.isDone()) {
```

Comment:
This is prone to race condition, future could complete just after `future.isDone()` is checked

Reply:
updated